firebase-installations/src/test/java/com/google/firebase/installations/remote/ResponseHandlerTest.java

```diff
@@ -0,0 +1,103 @@
+  private static final int RETRYABLE_RESPONSE_CODE = 500;
+  private static final long CURRENT_TIME_IN_MILLIS = 100000L;
+  private static final long NEXT_REQUEST_TIME_IN_MILLIS_LESSER_THAN_24H =
+      CURRENT_TIME_IN_MILLIS + TimeUnit.HOURS.toMillis(2);
+  private static final long NEXT_REQUEST_TIME_IN_MILLIS_GREATER_THAN_24H =
+      CURRENT_TIME_IN_MILLIS + TimeUnit.HOURS.toMillis(25);
+
+  @Mock private Utils mockUtils;
+  private ResponseHandler responseHandler;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    responseHandler = new ResponseHandler(mockUtils);
+  }
+
+  @Test
+  public void nonRetryableError_lessThan24Hr_doesNotRetry() {
+    when(mockUtils.currentTimeInMillis())
+        .thenReturn(CURRENT_TIME_IN_MILLIS, NEXT_REQUEST_TIME_IN_MILLIS_LESSER_THAN_24H);
+    responseHandler.setNextAllowedRequestTime(NON_RETRYABLE_RESPONSE_CODE);
+    assertFalse(responseHandler.isRequestAllowed());
+  }
+
+  @Test
+  public void nonRetryableError_greaterThan24Hr_retries() {
+    when(mockUtils.currentTimeInMillis())
+        .thenReturn(CURRENT_TIME_IN_MILLIS, NEXT_REQUEST_TIME_IN_MILLIS_GREATER_THAN_24H);
+    responseHandler.setNextAllowedRequestTime(NON_RETRYABLE_RESPONSE_CODE);
+    assertTrue(responseHandler.isRequestAllowed());
+  }
+
+  @Test
+  public void retryableError_greaterThanNextRequestTime_retries() {
+    when(mockUtils.getRandomMillis()).thenReturn(TimeUnit.MINUTES.toMillis(2));
+    when(mockUtils.currentTimeInMillis())
+        .thenReturn(CURRENT_TIME_IN_MILLIS, addMinutesToCurrentTime(5));
+    responseHandler.setNextAllowedRequestTime(RETRYABLE_RESPONSE_CODE);
+    assertTrue(responseHandler.isRequestAllowed());
+  }
+
+  @Test
+  public void retryableError_lesserThanNextRequestTime_doesNotRetry() {
+    when(mockUtils.getRandomMillis()).thenReturn(TimeUnit.MINUTES.toMillis(6));
+    when(mockUtils.currentTimeInMillis())
+        .thenReturn(CURRENT_TIME_IN_MILLIS, addMinutesToCurrentTime(5));
+    responseHandler.setNextAllowedRequestTime(RETRYABLE_RESPONSE_CODE);
+    assertFalse(responseHandler.isRequestAllowed());
+  }
+
+  @Test
+  public void retryableError_exponentialBackoff_retries() {
+    when(mockUtils.getRandomMillis()).thenReturn(TimeUnit.MINUTES.toMillis(5));
+    when(mockUtils.currentTimeInMillis())
+        .thenReturn(
+            CURRENT_TIME_IN_MILLIS,
+            addMinutesToCurrentTime(5),
+            addMinutesToCurrentTime(6),
+            addMinutesToCurrentTime(12));
+    responseHandler.setNextAllowedRequestTime(RETRYABLE_RESPONSE_CODE);
+    assertFalse(responseHandler.isRequestAllowed());
+    responseHandler.setNextAllowedRequestTime(RETRYABLE_RESPONSE_CODE);
+    assertTrue(responseHandler.isRequestAllowed());
+  }
+
+  // Adds specified minutes to the CURRENT_TIME_IN_MILLIS.
+  private long addMinutesToCurrentTime(int minutes) {
+    return CURRENT_TIME_IN_MILLIS + TimeUnit.MINUTES.toMillis(minutes);
+  }
```

Comment:
I think tests for the reset case will be beneficial. Tests that cover ResponseHandler.java lines 80-83 as well.